server/ladder_service/ladder_service.py

```diff
@@ -458,17 +482,111 @@ def on_match_found(
                     if name != queue.name
                 )
                 for queue_name in queue_names:
-                    self._cancel_search(player, queue_name)
+                    search = self._cancel_search(player, queue_name)
+                    if search is not None:
+                        original_searches[(search, queue_name)] = True
 
-                self._clear_search(player, queue.name)
-
-            asyncio.create_task(self.start_game(s1.players, s2.players, queue))
+            for player in players:
+                search = self._clear_search(player, queue.name)
+                if search is not None:
+                    original_searches[(search, queue.name)] = True
+
+            asyncio.create_task(
+                self.confirm_match(
+                    s1,
+                    s2,
+                    queue,
+                    list(reversed(original_searches))
+                )
+            )
         except Exception:
             self._logger.exception(
                 "Error processing match between searches %s, and %s",
                 s1, s2
             )
 
+    async def confirm_match(
+        self,
+        s1: Search,
+        s2: Search,
+        queue: MatchmakerQueue,
+        original_searches: list[tuple[Search, str]],
+    ):
+        self._logger.debug("original_searches %s", original_searches)
+        try:
+            all_players = s1.players + s2.players
+
+            offer = self.create_match_offer(all_players)
+            offer.write_broadcast_update()
+            await offer.wait_ready()
+
+            await self.start_game(s1.players, s2.players, queue)
+        except OfferTimeoutError:
+            unready_players = list(offer.get_unready_players())
+            assert unready_players, "Unready players should be non-empty if offer timed out"
+
+            self._logger.info(
+                "Match failed to start. Some players did not ready up in time: %s",
+                [player.login for player in unready_players]
+            )
+            self._cancel_match(all_players)
+
+            # Return any search that fully accepted the match back to the queue
+            for search, queue_name in original_searches:
+                self._logger.debug("%s, %s", search, queue_name)
+                search_players = search.players
+                search_unready_players = [
+                    player
+                    for player in unready_players
+                    if player in search_players
+                ]
+                if not search_unready_players:
+                    new_search = Search(
+                        players=search.players,
+                        # TODO: What start time do we want to use?
+                        start_time=search.start_time,
+                        rating_type=search.rating_type,
+                        on_matched=search.on_matched
+                    )
```

Comment:
Can we just reuse the same object? That way we preserve the failed matching attempts / threshold expansion which would give the auto requeue a real technical benefit over just having people rejoin the queue.

Reply:
Seems like yes